CEP 8 - Conda Release Schedule (#26)
conda-incubator/ceps: cep-releases.md

<table>
<tr><td> Title </td><td> Conda Release Schedule </td>
<tr><td> Status </td><td> Draft </td></tr>
<tr><td> Author(s) </td><td> Ken Odegard &lt;kodegard@anaconda.com&gt; </td></tr>
<tr><td> Created </td><td> May 20, 2022 </td></tr>
<tr><td> Updated </td><td> May 20, 2022 </td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

Describes a release cadence for all future `conda` versions starting with 5.0.0.

## Specification

> We propose a monthly release schedule where the tagging/release occurs during the week (per ISO 8601, Monday through Sunday) of the 15th.

Reviewer: ```suggestion
We propose a bi-monthly release schedule where the tagging/release occurs during the week (per ISO 8601, Monday through Sunday) of the 15th.
```

You mention a monthly release schedule in the alternatives below:

> 2. Follow a monthly schedule.
     - Rejected for being too fast for multi-user installs.

In a nod to our many different kinds of users we propose a deprecation policy (to be defined in a later CEP) that allows for a slower adoption rate. Users on a slower schedule would be encouraged to update once a quarter.

Both of these policies will be encapsulated in the version scheme:
- the major version will be an annual counter,
- the minor version will correlate with the quarter,
- the micro/patch version will generally correlate with the month within the quarter but will also be incremented if a bug fix/security patch needs an early release, and

> - this scheme will start with the July 2022 release of 5.0.0.

Reviewer 2: I like the general idea, but I'm wondering if we shouldn't just switch conda to using CalVer, which is essentially what's being proposed here. That would reduce any potential miscommunication caused by preserving something that looks like SemVer but really isn't. (E.g., whether conda 5.0.0 or 6.0.0 would be completely different from the current 4.x release series).

Author: Definitely an option as it would create a clear distinction between "old" and "new" conda.

Another thing to keep in mind with a more regular release cycle like this it's not infeasible for us to start planning out feature changes. Especially if we are seeking to allow a quarterly update schedule for slower-moving orgs. New features are added at the start of the quarter and the final release in the quarter is dedicated to cleanup/bugfixes.

Personally, I'd hesitate with switching entirely over to CalVer as it's generally a one-way path (the major version would do a significant one-way jump, 4 to 22 or 2022). It's hard to switch back again if that's desired for any reason.

An alternative would be to mirror Python's 2 → 3. We can do a one-time major version increment to 5 and drop the quarterly counter.

- 5.0.0: July 2022
- 5.0.1: August 2022 
- 5.0.2: September 2022
- ...
- 5.0.11: June 2022
- 5.1.0: July 2023
- 5.1.1: August 2023
- 5.1.2: September 2023
- ...

This is simpler but we'd lose the quarterly info but that could be instead solved with an LTS policy?

Author: Here are the current ideas and a new one. If we don't want to commit to a monthly release we should avoid a monthly counter as it'll result in versions jumping (e.g. 22.6 to 22.8 for bi-monthly releases) which could be confusing for users.

1. **YY.Q.MM**
    **\<year counter>.\<zero-based quarter counter>.\<zero-based month in quarter counter>**
    **<5+>.<0-3>.<0-3>**
    My original idea. Not in love with this idea as it made the version messy but nothing a little automation couldn't fix.
    - 5.0.0...5.0.1......5.1.0.........6.0.0
2. **YY.MM.MICRO**
    **\<year without century>.\<zero-based month counter>.\<zero-based release in month counter>**
    **<22+>.<0-11>.<0+>**
    @jezdez's proposed idea. This is _almost_ what we've already been using in https://github.com/conda/actions/releases. This would result in successful releases for a given month not having any micro/patch releases. Would also mean that the versions would jump if a release doesn't occur every month.
    - 22.0.0...22.0.1......22.1.0.........23.0.0
3. **YY.Q.MICRO**
    **\<year without century>.\<zero-based quarter counter>.\<zero-based release in quarter counter>**
    **<22+>.<0-3>.<0+>**
    This ends up looking very similar to @jezdez proposal but allows for slower releases, i.e. if we don't want to commit to monthly releases but want at least one per quarter.
    - 22.0.0...22.0.1......22.1.0.........23.0.0

Author: Ok adopted `YY.MM.MICRO` CalVer scheme on a bimonthly schedule.

> Both of these policies will be encapsulated in the version scheme:
> - the major version will be an annual counter,
> - the minor version will correlate with the quarter,
> - the micro/patch version will generally correlate with the month within the quarter but will also be incremented if a bug fix/security patch needs an early release, and
> - this scheme will start with the July 2022 release of 5.0.0.

Reviewer: ```suggestion
Both of these policies will be encapsulated in the version scheme:
- the major version will be an annual counter,
- the minor version will be a running counter of releases within the year,
- the micro/patch will be incremented if a bug fix/security patch needs an early release
- this scheme will start with the July 2022 release of 22.0.0.
```

> This will result in the following release schedule:
> - 5.0.0: 2022-07-14
> - 5.0.1: 2022-08-18 (5 weeks of development)
> - 5.0.2: 2022-09-15 (4 weeks)
> - 5.1.0: 2022-10-13 (4 weeks)
> - 5.1.1: 2022-11-17 (5 weeks)
> - 5.1.2: 2022-12-15 (4 weeks)
> - 5.2.0: 2023-01-12 (4 weeks)
> - 5.2.1: 2023-02-16 (5 weeks)
> - 5.2.2: 2023-03-16 (4 weeks)
> - 5.3.0: 2023-04-13 (4 weeks)
> - 5.3.1: 2023-05-18 (5 weeks)
> - 5.3.2: 2023-06-15 (4 weeks)
> - 6.0.0: 2023-07-13 (4 weeks)

Reviewer: ```suggestion
This will result in the following release schedule (where the bugfix releases are ad-hoc and just here just for presentation purposes):
- 22.0.0: 2022-07-14 (first regular release)
- 22.0.1: 2022-07-18 (bugfix release)
- 22.1.0: 2022-09-15 (second regular release)
- 22.1.1: 2022-10-01 (bugfix release)
- 22.2.0: 2022-11-17 (third regular release)
- 22.2.1: 2022-12-15 (bugfix release)
- 23.0.0: 2023-01-12 (first regular release)
- 23.0.1: 2023-01-16 (bugfix release)
- 23.0.2: 2023-01-19 (bugfix release 2)
- 23.1.0: 2023-03-16 (second regular release)
- 23.1.0: 2023-04-13 (bugfix release)
- 23.2.0: 2023-05-18 (third regular release)
- 23.2.1: 2023-05-23 (bugfix release)
- 23.2.0: 2023-05-18 (third regular release)
...
```

<table>
<tr><td> Title </td><td> Conda Release Schedule </td>
<tr><td> Status </td><td> Draft </td></tr>

> <tr><td> Author(s) </td><td> Ken Odegard &lt;kodegard@anaconda.com&gt; </td></tr>

Reviewer: ```suggestion
<tr><td> Author(s) </td>
<td> Ken Odegard &lt;kodegard@anaconda.com&gt;, Jannis Leidel &lt;jleidel@anaconda.com&gt; </td></tr>
```

<tr><td> Created </td><td> May 20, 2022 </td></tr>
<tr><td> Updated </td><td> July 20, 2022 </td></tr>

> <tr><td> Discussion </td><td> NA </td></tr>

Reviewer: ```suggestion
<tr><td> Discussion </td><td> https://github.com/conda-incubator/ceps/pull/26 </td></tr>
```

<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

Describes a release cadence for all future `conda` versions starting with `22.7.0`.

## Specification

We propose regularly scheduled bimonthly (every two months) releases where the tagging/release occurs during the week (per ISO 8601, Monday through Thursday) of second Monday of the month.

> In a nod to our many different kinds of users we will propose a deprecation policy (to be defined in a later CEP) that allows for a slower adoption rate (i.e. users could update every 3-4 months instead).

Reviewer 3: ```suggestion
In a nod to our many different kinds of users, we will also propose a deprecation policy (to be defined in a later CEP) that allows for a slower adoption rate (i.e. users could update every 3-4 months instead).
```

> These objectives will be aided in the adoption of [CalVer](https://calver.org/):

Reviewer 3: ```suggestion
To accomplish better, more predictable versioning, we will adopt [CalVer](https://calver.org/):
```

- `YY`: the major version will be the shortened year (22+)
- `MM`: the minor version will be the shortened month (1-12)
- `MICRO`: the micro/patch version is reset to zero every month and incremented for every additional release that occurs during that month (0+)

This scheme will start with the July 2022 release of 22.7.0.

This will result in the following regular release schedule:

| Version | Release Week |
|---|---|
| 22.7.0 | 2022-07-11 |
| 22.9.0 | 2022-09-12 |
| 22.11.0 | 2022-11-14 |
| 23.1.0 | 2023-01-09 |
| 23.3.0 | 2023-03-13 |
| 23.5.0 | 2023-05-08 |
| 23.7.0 | 2023-07-10 |
| 23.9.0 | 2023-09-11 |
| ... | ... |

> **Note**

> > Despite following a bimonthly release schedule we will permit releases to occur at any time between regular releases for hotfixes, security releases, or other high priority changes that require immediate release.

Reviewer 3: ```suggestion
> Despite following a bimonthly release schedule, we will permit releases to occur at any time between regular releases for hot-fixes, security releases, or other high priority changes that require immediate release.
```

Author: I agree that "hotfix" isn't officially a word, but within tech, it's supposed to be unhyphenated (per [Google](https://www.google.com/search?q=define+hotfix) and [Wikipedia](https://en.wikipedia.org/wiki/Hotfix)).

> To distinguish between the bimonthly release schedule and other optional releases the following release types are defined:

Reviewer 3: ```suggestion
To distinguish between the bimonthly release schedule and other optional releases, the following release types are defined:
```

> - **regular release**: this is the regularly schedule bimonthly release

Reviewer 3: ```suggestion
- **Regular release**: this is the regularly schedule bimonthly release
```

> - **optional release**: releases that may occur on the alternating months from regular releases

Reviewer 3: ```suggestion
- **Optional release**: releases that may occur on the alternating months from regular releases
```

> - **hotfix release**: extra releases that may occur during any month to patch an earlier release of the same month (whether a regular or optional release)

Reviewer 3: ```suggestion
- **Hotfix release**: extra releases that may occur during any month to patch an earlier release of the same month (whether a regular or optional release)
```

> So it's entirely feasible to see the following releases:

Reviewer 3: ```suggestion
So, it's entirely feasible to see the following releases:
```

| Version | Release Type |
|---|---|
| 22.7.0 | regular |
| 22.7.1 | hotfix |
| 22.7.2 | hotfix |
| 22.8.0 | optional |
| 22.8.1 | hotfix |
| 22.9.0 | regular |
| 22.10.0 | optional |
| 22.10.1 | hotfix |
| ... | ... |

## Motivation

> Remove ambiguity/maintainer guesswork of when and what warrants a release.

Reviewer 3: ```suggestion
Our goal with this CEP is removing ambiguity/maintainer guesswork of when and what warrants a release.
```

## Backwards Compatibility

Adopting this release schedule does not break any existing processes or schemes.

We will entirely sidestep and avoid any confusion regarding years of rumored changes slated to be included in a `conda` 5 release since we won't even attain a 5 release with this new scheme. This is not to say that these desired changes wont occur, but rather that its unrealistic for all of these changes to be successfully (and safely) rolled out in a single release.

> 

Reviewer 3: What about breaking tools that require a certain version format? It seems like that won't be the case here, but it may be nice to briefly mention that too (i.e. that CalVer is compatible with SemVer).

Author: Added

Author: Added
